src/edg.rs

```diff
@@ -156,97 +189,167 @@ impl<B: Broker<V> + Debug, G: ExtendedDependencyGraph<V> + Send + Sync + Debug,
+                        );
+                        self.process_hyper_edge(edge, weight)
                     }
+                    Err(err) => panic!(
+                        "Receiving from hyper edge waiting channel failed with: {}",
+                        err
+                    ),
                 }
-                // Received more work
-                i if i == oper_msg => {
-                    match oper.recv(&msg_rx) {
-                        // Alg 1, Line 5-9
-                        Ok(msg) => {
-                            match msg {
-                                // Alg 1, Line 6
-                                Message::HYPER(edge, weight) => {
-                                    self.process_hyper_edge(edge, weight)
+            } else if !negation_rx.is_empty() {
+                match negation_rx.recv() {
+                    Ok((edge, weight)) => {
+                        let _guard = span!(
+                            Level::TRACE,
+                            "worker receive negation-edge",
+                            worker_id = self.id,
+                            ?edge,
+                            ?weight
+                        );
+                        match self.assignment.get(&edge.target) {
+                            None => self.process_negation_edge(edge.clone(), weight),
+                            // Alg 1, Line 7
+                            Some(assignment) => match assignment {
+                                VertexAssignment::FALSE => {
+                                    self.process_negation_edge(edge.clone(), weight)
                                 }
-                                // Alg 1, Line 7
-                                Message::NEGATION(edge, weight) => {
-                                    self.process_negation_edge(edge, weight)
+                                VertexAssignment::TRUE => {
+                                    self.process_negation_edge(edge.clone(), weight)
                                 }
-                                // Alg 1, Line 8
-                                Message::REQUEST {
-                                    vertex,
-                                    worker_id,
-                                    weight,
-                                } => self.process_request(&vertex, worker_id, weight),
-                                // Alg 1, Line 9
-                                Message::ANSWER {
-                                    vertex,
-                                    assignment,
-                                    weight,
-                                } => self.process_answer(&vertex, assignment, weight),
-                            }
+                                // In case of undecided the edge is only processed after 10 iterations
+                                VertexAssignment::UNDECIDED => {
+                                    if self.counter == 9 {
+                                        self.counter = 0;
```

Comment:
I think we should reset the counter whenever we do something that is not a negation edge too.